088-rebalance-progress-status.md

```diff
@@ -0,0 +1,334 @@
+    monitoredPartitionsPercentage: 100
+    numIntraBrokerReplicaMovements: 0
+    numLeaderMovements: 16
+    numReplicaMovements: 0
+    onDemandBalancednessScoreAfter: 95.4347095948149
+    onDemandBalancednessScoreBefore: 89.4347095948149
+    provisionRecommendation: ""
+    provisionStatus: RIGHT_SIZED
+    recentWindows: 1
+  progress: [1]
+    rebalanceProgressConfigMap: my-rebalance-progress [2]
+```
+[1] The `progress` section will be visible during the KafkaRebalance `Rebalancing` and  `Stopped` states.
+
+[2] The `ConfigMap` containing information related to the ongoing partition rebalance, generated with the name "<kafka_rebalance_resource_name>-progress".
+
+In the `ConfigMap`, we will include the following fields:
+- **estimatedTimeToCompletion**: The estimated amount time it will take in minutes until partition rebalance is complete.
+- **percentageDataMovementComplete**: The percentage of the data movement of the partition rebalance that is completed e.g. values in the range [0-100]%
+- **executorState**: The “non-verbose” JSON payload from the `/kafkacruisecontrol/state?substates=executor` endpoint.
+
+The progress `ConfigMap` of an inter-broker partition rebalance would look like the following:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-rebalance-progress
+  …
+data:
+  estimatedTimeToCompletion: 5m [1]
+  percentageDataMovementComplete: 80% [2]
+  executorState: [3]
+  {
+   "abortingPartitions":0,
+   "averageConcurrentPartitionMovementsPerBroker":5,
+   "finishedDataMovement":0,
+   "maximumConcurrentPartitionMovementsPerBroker":5,
+   "minimumConcurrentPartitionMovementsPerBroker":5,
+   "numFinishedPartitionMovements":0,
+   "numInProgressPartitionMovements":0,
+   "numPendingPartitionMovements":20,
+   "numTotalPartitionMovements":20,
+   "state":"INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS",
+   "totalDataToMove":0,
+   "triggeredSelfHealingTaskId":"",
+   "triggeredTaskReason":"No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
+   "triggeredUserTaskId":"0230d401-6a36-430e-9858-fac8f2edde93"
+  }
+```
+[1] The estimated time it will take the rebalance to complete based on the average rate of data transfer.
+
+[2] The percentage complete of the ongoing rebalance in the range [0-100]%
+
+[3] The “non-verbose” JSON payload from [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint.
+
+### Field values per `KafkaRebalance` State
+
+For the initial implementation, we will focus on providing the progress information in the following `KafkaRebalance` states:
+
+- `Rebalancing`
+- `Stopped`
+
+These are the states where this progress information will be able to be most accurately calculated and most useful for users. 
+We could provide this information for other states as well, such as the `ProposalReady` and `Ready` states, but it is not completely necessary, nor is it trivial. 
```

Comment:
I still argue that having a lower bound for the estimated time of a `KafkaRebalance` in the `ProposalReady` state, based on the proposed data to move and the throttle settings would be a useful thing to have. It would obviously not be accurate as there is not disk throughput available. 

But it is still a useful guide and helps users gauge the impact of a proposed rebalance, which data-to-move values alone don't give.